crypto/kyber/p_kyber_test.cc

```diff
@@ -4,280 +4,463 @@
 #include <openssl/evp.h>
 #include <openssl/mem.h>
 
-#include "../crypto/evp_extra/internal.h"
+#include "../test/file_test.h"
+#include "../test/test_util.h"
+#include "../evp_extra/internal.h"
 #include "../fipsmodule/evp/internal.h"
+#include "../rand_extra/pq_custom_randombytes.h"
 #include "../internal.h"
 #include "kem_kyber.h"
 
-TEST(Kyber512Test, KeyGeneration) {
-  EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, nullptr);
-  ASSERT_NE(kyber_pkey_ctx, nullptr);
+static uint8_t* get_pub_from_kyber_key(EVP_PKEY *pkey) {
+  if (EVP_PKEY_id(pkey) == EVP_PKEY_KYBER512) {
+    KYBER512_KEY *key = (KYBER512_KEY *)(pkey->pkey.ptr);
+    return key->pub;
+  } else if (EVP_PKEY_id(pkey) == EVP_PKEY_KYBER768) {
+    KYBER768_KEY *key = (KYBER768_KEY *)(pkey->pkey.ptr);
+    return key->pub;
+  }
+  return nullptr;
+}
+
+static uint8_t* get_priv_from_kyber_key(EVP_PKEY *pkey) {
+  if (EVP_PKEY_id(pkey) == EVP_PKEY_KYBER512) {
+    KYBER512_KEY *key = (KYBER512_KEY *)(pkey->pkey.ptr);
+    return key->priv;
+  } else if (EVP_PKEY_id(pkey) == EVP_PKEY_KYBER768) {
+    KYBER768_KEY *key = (KYBER768_KEY *)(pkey->pkey.ptr);
+    return key->priv;
+  }
+  return nullptr;
+}
+
+static uint8_t get_has_private_from_kyber_key(EVP_PKEY *pkey) {
+  if (EVP_PKEY_id(pkey) == EVP_PKEY_KYBER512) {
+    KYBER512_KEY *key = (KYBER512_KEY *)(pkey->pkey.ptr);
+    return key->has_private;
+  } else if (EVP_PKEY_id(pkey) == EVP_PKEY_KYBER768) {
+    KYBER768_KEY *key = (KYBER768_KEY *)(pkey->pkey.ptr);
+    return key->has_private;
+  }
+  return 0;
+}
+
+static size_t kyber_public_key_size(int pkey_id) {
+  if (pkey_id == EVP_PKEY_KYBER512) {
+    return KYBER512_PUBLIC_KEY_BYTES;
+  } else if (pkey_id == EVP_PKEY_KYBER768) {
+    return KYBER768_PUBLIC_KEY_BYTES;
+  }
+  return 0;
+}
+
+static size_t kyber_secret_key_size(int pkey_id) {
+  if (pkey_id == EVP_PKEY_KYBER512) {
+    return KYBER512_SECRET_KEY_BYTES;
+  } else if (pkey_id == EVP_PKEY_KYBER768) {
+    return KYBER768_SECRET_KEY_BYTES;
+  }
+  return 0;
+}
+
+static size_t kyber_shared_secret_size(int pkey_id) {
+  if (pkey_id == EVP_PKEY_KYBER512) {
+    return KYBER512_SHARED_SECRET_BYTES;
+  } else if (pkey_id == EVP_PKEY_KYBER768) {
+    return KYBER768_SHARED_SECRET_BYTES;
+  }
+  return 0;
+}
+
+static size_t kyber_ciphertext_size(int pkey_id) {
+  if (pkey_id == EVP_PKEY_KYBER512) {
+    return KYBER512_CIPHERTEXT_BYTES;
+  } else if (pkey_id == EVP_PKEY_KYBER768) {
+    return KYBER768_CIPHERTEXT_BYTES;
+  }
+  return 0;
+}
 
-  EVP_PKEY *kyber_pkey = EVP_PKEY_new();
-  ASSERT_NE(kyber_pkey, nullptr);
+static const int pkeyIDsToTest[2] = { EVP_PKEY_KYBER512,
+                                      EVP_PKEY_KYBER768 };
 
-  EXPECT_TRUE(EVP_PKEY_keygen_init(kyber_pkey_ctx));
-  EXPECT_TRUE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey));
-  ASSERT_NE(kyber_pkey->pkey.ptr, nullptr);
+TEST(KyberTest, KeyGeneration) {
+  for (auto pkey_id : pkeyIDsToTest) {
```

Comment:
Did you consider using a [parameterized gtest](https://github.com/google/googletest/blob/main/docs/advanced.md#value-parameterized-tests) like aead_test.c? It accomplishes basically the same thing, but you get nicer test output:
* As is we would see "KeyGeneration failed assert on line 90", and not be able to tell which kyber level failed
* With paramertized tests we would get "KeyGeneration/Kyber512 failed assert on line 90"

Reply:
neat :) didn't know about this type of tests, I'll reimplement.